src/coreclr/jit/rationalize.cpp

```diff
@@ -369,6 +369,13 @@ void Rationalizer::RewriteAssignment(LIR::Use& use)
     GenTreeOp* assignment = use.Def()->AsOp();
     assert(assignment->OperGet() == GT_ASG);
 
+    if (comp->fgIsSafeToRemoveCastOnAssignment(assignment))
+    {
+        GenTreeCast* cast = assignment->gtGetOp2()->AsCast();
+        assignment->gtOp2 = cast->CastOp();
+        BlockRange().Remove(cast);
+    }
```

Comment:
Why move it to rationalization? We should keep it in morph, it is quite weird to have transforms in rationalization (the mod one os an exception, it should not be the norm).

Reply:
I'm testing it out. I can't seem to get the regression test to pass when it's in morph; I thought it was working, but it was not. I forgot to set my envars :(

Reply:
But it looks like other tests break when it's in rationalization. I can easily put it back in morph; I've made the helper function.

Reply:
Hmm, I think there is still an assertion prop issue here like the one @SingleAccretion talked about, even with the cast fixed so that the IR is well-typed. This sequence is quite suspicious:
```scala
The assignment [000010] using V00 removes: Constant Assertion: V00 == 246
GenTreeNode creates assertion:
               [000010] -A---------                         ▌  ASG       ubyte 
In BB01 New Local Constant Assertion: V00 == -1, index = #02

fgMorphTree BB01, STMT00001 (after)
               [000010] -A---+-----                         ▌  ASG       ubyte 
               [000009] D----+-N---                         ├──▌  LCL_VAR   ubyte  V00 arg0         
               [000023] -A---+-----                         └──▌  COMMA     int   
               [000017] -A---+-----                            ├──▌  ASG       int   
               [000016] D----+-N---                            │  ├──▌  LCL_VAR   int    V02 tmp1         
               [000006] -----+-----                            │  └──▌  CNS_INT   int    247
               [000022] -----+-----                            └──▌  CNS_INT   int    -1

fgMorphTree BB01, STMT00002 (before)
               [000013] -----------                         ▌  RETURN    long  
               [000012] ---------U-                         └──▌  CAST      long <- ulong <- uint
               [000011] -----------                            └──▌  LCL_VAR   ubyte  V00 arg0         

Assertion prop in BB01:
Constant Assertion: V00 == -1, index = #02
               [000011] -----------                         ▌  CNS_INT   int    -1

Folding long operator with constant nodes into a constant:
               [000012] ---------U-                         ▌  CAST      long <- ulong <- uint
               [000011] -----+-----                         └──▌  CNS_INT   int    -1
Bashed to long constant:
               [000012] -----------                         ▌  CNS_INT   long   0xFFFFFFFF

fgMorphTree BB01, STMT00002 (after)
               [000013] -----+-----                         ▌  RETURN    long  
               [000012] -----+-----                         └──▌  CNS_INT   long   0xFFFFFFFF
```
The assertion created is wrong.

Simple repro with inline IL:
```csharp
public static void Main()
{
    System.Console.WriteLine(M50(0));
}

[MethodImpl(MethodImplOptions.NoInlining)]
public static long M50(byte arg0)
{
    IL.Emit.Ldc_I4(-1);
    IL.Emit.Starg(0);
    IL.Emit.Ldarg(0);
    IL.Emit.Conv_I8();
    return IL.Return<long>();
}
```
Expected: 255
Actual: -1 in release, 255 in debug

Reply:
The above is on your commit [8e56f46](https://github.com/dotnet/runtime/pull/70518/commits/8e56f4669ff216945dbecb4836cc64de993c7346).

Reply:
I think both the assertion created is wrong and it also seems wrong that we allow the replacement of a ubyte typed `LCL_VAR` with a non zero-extended value.

Reply:
I see. So, there are two separate issues. One with the narrow cast, and the other in assertion prop. 

Having the check for VM type equality does fix one the issues. Perhaps I should just scope this PR to just the narrow cast fix.